chain/network/src/accounts_data/tests.rs

```diff
@@ -0,0 +1,207 @@
+    let a1 = signers[1].make_account_data(rng, now);
+    let mut a2_too_large = signers[2].make_account_data(rng, now);
+    *a2_too_large.payload_mut() =
+        (0..crate::network_protocol::MAX_ACCOUNT_DATA_SIZE_BYTES + 1).map(|_| 17).collect();
+
+    // too large payload => DataTooLarge
+    let res = cache
+        .clone()
+        .insert(vec![
+            a0.clone(),
+            a1.clone(),
+            a2_too_large.clone(), // invalid entry => DataTooLarge
+        ])
+        .await;
+    assert_eq!(Some(Error::DataTooLarge), res.1);
+    // Partial update is allowed, in case an error is encountered.
+    assert!([&a0, &a1].as_set().is_superset(&res.0.as_set()));
+    // Partial update should match the state.
+    assert_eq!(res.0.as_set(), cache.dump().as_set());
+}
+
+#[tokio::test]
+async fn invalid_signature() {
+    let mut rng = make_rng(2947294234);
+    let rng = &mut rng;
+    let clock = time::FakeClock::default();
+    let now = clock.now_utc();
+
+    let signers = make_signers(rng, 3);
+    let e = make_account_keys(&signers);
+
+    let cache = Arc::new(Cache::new());
+    cache.set_keys(e);
+    let a0 = signers[0].make_account_data(rng, now);
+    let mut a1 = signers[1].make_account_data(rng, now);
+    let mut a2_invalid_sig = signers[2].make_account_data(rng, now);
+    *a2_invalid_sig.signature_mut() = a1.signature_mut().clone();
+
+    // invalid signature => InvalidSignature
+    let res = cache
+        .clone()
+        .insert(vec![
+            a0.clone(),
+            a1.clone(),
+            a2_invalid_sig.clone(), // invalid entry => DataTooLarge
+        ])
+        .await;
+    assert_eq!(Some(Error::InvalidSignature), res.1);
+    // Partial update is allowed, in case an error is encountered.
+    assert!([&a0, &a1].as_set().is_superset(&res.0.as_set()));
+    // Partial update should match the state.
+    assert_eq!(res.0.as_set(), cache.dump().as_set());
+}
+
+#[tokio::test]
+async fn single_account_multiple_data() {
+    let mut rng = make_rng(2947294234);
+    let rng = &mut rng;
+    let clock = time::FakeClock::default();
+    let now = clock.now_utc();
+
+    let signers = make_signers(rng, 3);
+    let e = make_account_keys(&signers);
+
+    let cache = Arc::new(Cache::new());
+    cache.set_keys(e);
+    let a0 = signers[0].make_account_data(rng, now);
+    let a1 = signers[1].make_account_data(rng, now);
+    let a2old = signers[2].make_account_data(rng, now);
+    let a2new = signers[2].make_account_data(rng, now + time::Duration::seconds(1));
+
+    // 2 entries for the same (epoch_id,account_id) => SingleAccountMultipleData
+    let res =
+        cache.clone().insert(vec![a0.clone(), a1.clone(), a2old.clone(), a2new.clone()]).await;
+    assert_eq!(Some(Error::SingleAccountMultipleData), res.1);
+    // Partial update is allowed, in case an error is encountered.
+    assert!([&a0, &a1, &a2old, &a2new].as_set().is_superset(&res.0.as_set()));
```

Comment:
Rather than adding `AsSet` trait, I think it'll be more valuable to add an `#[track_caller] fn assert_super_set` function to emit a useful error message her.

Reply:
I assume that you are fine with assert_eq(x.as_set(),y.as_set()) and the concern is about just the error message for is_superset. I'll try to find a way to pretty print the diff for that cases.

Reply:
done